Fix issue 20245 - address of ref should be scope

dkorpel opened the pull request.

dkorpel committed:
fix issue 20245 - address of ref should be scope

MoonlightSentinel commented on test/fail_compilation/test20245.d:
```diff
@@ -0,0 +1,44 @@
+}
+
+@safe int** foo(ref scope int* x) {
+    int** a = &x;
+    return a;
```
This error is inconsistent to the other test cases, `a` should be inferred as `scope`. Seems to be yet another weird special case based on `hasIndirections`?

dkorpel replied:
There's two checks: first, in expressionsem.d there's `checkAddressVar` to check whether it is possible at all to take the address safely. This fails because it's taking the address of a `scope` variable with pointers, and there's only one layer of `scope`. This test case is taken from Walter's issue description:
> The following should not compile:
> (...)
> because there's not way to attach `scope` to anything but the head, i.e. no scope pointers to scope pointers to int.

The second check is in escape.d where all the `scope` / `return` is inferred and escaping `scope` variables are detected, which is where the "scope variable `a` may not be returned" comes from.

MoonlightSentinel replied:
Right, DIP1000 falls apart once there's multiple layers of indirections. Got tricked again by the kinda misleading error message

dkorpel committed:
allow escaping &ref without dip1000

dkorpel commented:
After I fixed the buildkite failure of vibe-d I was afraid all red packages had to update the version of their vibe-d dependency, but then I realized it should only fail with dip1000, so I added a dip1000 check to and it's green now. Ready for review.

Geod24 commented:
But Vibe.d doesn't compile with DIP1000 ?

dkorpel commented:
> But Vibe.d doesn't compile with DIP1000 ?

Exactly, so the initial buildkite failure of vibe-d and dependents indicated that I needed to add the `(global.params.useDIP1000 != FeatureState.enabled && v.isReference())` which I did, so it only affects code compiled with dip1000.
